docker/resource_docker_service.go

```diff
@@ -0,0 +1,490 @@
+			},
+			"env": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Set:      schema.HashString,
+			},
+			"hosts": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				ForceNew: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"ip": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+							ForceNew: true,
+						},
+
+						"host": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+							ForceNew: true,
+						},
+					},
+				},
+			},
+			"network_mode": &schema.Schema{
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validateStringMatchesPattern(`^(vip|dnsrr)$`),
+			},
+			"networks": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Set:      schema.HashString,
+			},
+			"mounts": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"target": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"source": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"type": &schema.Schema{
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validateStringMatchesPattern(`^(bind|volume|tmpf)$`),
+						},
+						"consistency": &schema.Schema{
+							Type:         schema.TypeString,
+							Optional:     true,
+							ValidateFunc: validateStringMatchesPattern(`^(default|consistent|cached|delegated)$`),
+						},
+						"read_only": &schema.Schema{
+							Type:     schema.TypeBool,
+							Optional: true,
+						},
+						"bind_propagation": &schema.Schema{
+							Type:         schema.TypeString,
+							Optional:     true,
+							ValidateFunc: validateStringMatchesPattern(`^(private|rprivate|shared|rshared|slave|rslave)$`),
+						},
+						"volume_no_copy": &schema.Schema{
+							Type:     schema.TypeBool,
+							Optional: true,
+						},
+						"volume_labels": &schema.Schema{
+							Type:     schema.TypeMap,
+							Optional: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+						},
+						"volume_driver_name": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"volume_driver_options": &schema.Schema{
+							Type:     schema.TypeSet,
+							Optional: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+							Set:      schema.HashString,
+						},
+						"tmpfs_size_bytes": &schema.Schema{
+							Type:     schema.TypeInt,
+							Optional: true,
+						},
+						"tmpfs_mode": &schema.Schema{
+							Type:     schema.TypeInt,
+							Optional: true,
+						},
+					},
+				},
+			},
+
+			"configs": &schema.Schema{
```

Comment:
Are `configs` and `secrets` here the same things as `resource_docker_config` and `resource_docker_secret`? How do users use these resources together? The `service` documentation has users supply a block with `config.id`, `config.name`, and then the data being a path to a file, but not `config.data`? 

If they are meant to be used together, I think we may want to simply have `configs` and `secrets` be a `TypeSet` of IDs, if possible. Thoughts?

Reply:
No, they are not the same. `configs` and `secrets` is base64 encoded data which is later mounted into the container of the service at the given `file_path`. I tried to explain it properly in the docs and added extensive tests for these use cases but maybe I'd link the official documentation which can be found [here](https://docs.docker.com/engine/swarm/configs). Yes?